In [2]:
###############################################################################
# Date:        Nov. 1st 2019
# Name:      Peiying Lyu 
#ID:              8109407016
# Course:    INF 552, HW5
################################################################################

# 1 Multi-class and Multi-Label Classification Using Support Vector Machines 

## (a) download dataset and 70% as train data

In [1]:
import numpy as np
import pandas as pd
import csv
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings('ignore')


df = pd.read_csv("D:/552 Machine Learning for Data Science/homework/Homework5/Anuran Calls (MFCCs)/Frogs_MFCCs.csv",
                                               sep = ',', header = 0)

print(df.shape)
train_df, test_df = train_test_split(df, train_size=0.7, shuffle=True, random_state=0)

print(train_df.shape)
print(test_df.shape)

(7195, 26)
(5036, 26)
(2159, 26)


In [2]:
train_df

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  MFCCs_ 4  MFCCs_ 5  MFCCs_ 6  MFCCs_ 7  \
854   1.000000 -0.110506  0.296575  0.572484  0.106014  0.071651  0.026353   
4279  1.000000  0.220624  0.242714  0.687612  0.278120  0.048901 -0.175033   
6530  1.000000  0.537389  0.411925  0.220964  0.082649  0.116707  0.030271   
4445  1.000000  0.113471  0.141275  0.510700  0.171047  0.004309 -0.152609   
4218  1.000000  0.240949  0.216480  0.461523  0.117414 -0.062445 -0.130187   
1606  1.000000  0.140276  0.044974  0.547443  0.188967  0.026188 -0.175410   
480   1.000000  0.453858  0.626527  0.615896 -0.200647 -0.146663  0.450525   
3184  1.000000  0.123081  0.005561  0.413929  0.221739  0.114452 -0.039277   
3509  1.000000  0.713326  0.386898  0.478786  0.126294  0.001281 -0.019061   
2102  1.000000  0.493951  0.141637  0.439138  0.168985  0.071982 -0.040581   
446   1.000000  0.452371  0.606722  0.473059 -0.253796 -0.101704  0.401613   
4480  1.000000  0.116445  0.174462  0.619721  0.267264  0.018992 -0.204976   
1273  1.000000  0.486583  0.430116  0.523118  0.082630  0.115019 -0.020961   
552   1.000000  0.470385  0.688692  0.537746 -0.236105 -0.080203  0.433915   
549   1.000000  0.506183  0.710849  0.638431 -0.152501 -0.025995  0.460799   
6572  0.502739  0.961007  1.000000  0.573768  0.025829  0.455644  0.073500   
5506  1.000000  0.554986  0.472957  0.183668  0.022795  0.226007  0.046965   
2920  1.000000  0.290938  0.204431  0.540228  0.167556  0.062985 -0.036107   
6263  1.000000  0.534694  0.497449  0.292324 -0.023594  0.169344  0.101634   
4809  1.000000  0.150568  0.101737  0.481503  0.288187  0.109403  0.012022   
1212  1.000000  0.411596  0.298464  0.425783  0.117526  0.085654  0.144123   
2089  1.000000  0.330732  0.103805  0.442383  0.172698  0.110279  0.046838   
6720  1.000000  0.092693  0.450444  0.320916 -0.146515  0.307638  0.237495   
5955  1.000000  0.021701  0.410478  0.399667  0.171750  0.213241  0.134878   
6405  1.000000  0.567712  0.416835  0.257801  0.054688  0.171091  0.051874   
4542  1.000000  0.152593  0.235992  0.624108  0.176880  0.000336 -0.096379   
1231  1.000000  0.190594  0.387170  0.707107  0.168198  0.009813  0.025088   
3425  1.000000  0.251102  0.116712  0.632898  0.338071  0.106538 -0.111474   
5663  1.000000  0.260534  0.316417  0.250214  0.117294  0.142602  0.089499   
5970  1.000000  0.143791  0.319142  0.308636  0.155383  0.198975  0.115934   
...        ...       ...       ...       ...       ...       ...       ...   
2046  1.000000  0.269340  0.181261  0.651894  0.232370  0.038751 -0.048521   
4851  1.000000  0.074362  0.263197  0.615827  0.136339 -0.079374  0.105150   
5072  1.000000  0.667976  0.622274  0.329584 -0.045869  0.037852 -0.085010   
2163  1.000000  0.024757  0.058893  0.653633  0.345012  0.036526 -0.183545   
6036  1.000000 -0.145014  0.525495  0.455897  0.149029  0.350484  0.044916   
6921  0.875556  1.000000  0.924700 -0.162617  0.254338  0.077314 -0.143668   
6216  1.000000  0.271352  0.313735  0.223955  0.133285  0.189383  0.083144   
2893  1.000000  0.725058  0.411548  0.881384  0.090878 -0.040575 -0.167982   
537   1.000000  0.500723  0.646848  0.538545 -0.220942 -0.040400  0.372114   
1701  1.000000  0.286636  0.081918  0.555118  0.247951  0.109042 -0.096193   
2897  1.000000  0.911240  0.524185  0.746059  0.033140 -0.019817 -0.020242   
2222  1.000000  0.057382  0.190249  0.538383  0.092533 -0.030746 -0.151134   
2135  1.000000  0.105988  0.265259  0.725896  0.133375 -0.053839 -0.075920   
2599  1.000000  0.348866  0.428769  0.662567  0.161041  0.031640 -0.065532   
705   1.000000  0.235306 -0.067987  0.166875  0.262280  0.255858  0.127852   
6458  1.000000  0.535783  0.421344  0.256011  0.035078  0.133310  0.052785   
3468  1.000000  0.228404  0.171016  0.547707  0.244747  0.128172 -0.083801   
6744  1.000000  0.109073  0.599177  0.411636 -0.241895  0.251065  0.320763   
6743  1.000000  0.072710  0.398125  0.414401 -0.020922  0.311403  0.294552   
5924  1.000000  0.172811  0.

## (b) to solve a multi-class and multi-label problem.
#### One of the most important approaches to multi-class classification is to train a classification <b>for each label</b>.

###  i .exact match& hamming score/ loss methods

#### Ans:
#### Hamming Loss
The fraction of the wrong labels to the total number of labels
#### Exact Match
It is the most strict metric, indicating the percentage of samples that have all their labels classified correctly.

### Preprocessing Data

In [3]:
from sklearn.metrics import accuracy_score
from sklearn.svm import SVC
from sklearn.model_selection import StratifiedKFold
from sklearn.preprocessing import StandardScaler

#train
feature_train = train_df.iloc[:,:22]
label_train = train_df.iloc[:,-4:-1]

#test
feature_test = test_df.iloc[:,:22]
label_test = test_df.iloc[:, -4:-1]

#scaler
scaler=StandardScaler()
fea_train_scale = scaler.fit_transform(feature_train)
fea_test_scale = scaler.fit_transform(feature_test)

print(feature_train.shape)
print(label_train.shape)

print(feature_test.shape)
print(label_test.shape)

categories  = ['Family', 'Genus', 'Species']

(5036, 22)
(5036, 3)
(2159, 22)
(2159, 3)


In [36]:
C = np.logspace(-3,6,num=10)
G = 0.1*np.arange(1,21,1)

print(C)
print(G)



[1.e-03 1.e-02 1.e-01 1.e+00 1.e+01 1.e+02 1.e+03 1.e+04 1.e+05 1.e+06]
[0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7 1.8
 1.9 2. ]


### ii. SVM using Gaussian Kernel / no Standard

#### find c & gamma:

In [54]:
import numpy as np
from sklearn.model_selection import StratifiedKFold
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

'''
https://jakevdp.github.io/PythonDataScienceHandbook/05.07-support-vector-machines.html
'''
#train
feature_train = train_df.iloc[:,:22]
label_train = train_df.iloc[:,-4:-1]

#test
feature_test = test_df.iloc[:,:22]
label_test = test_df.iloc[:, -4:-1]

#scaler
scaler=StandardScaler()
fea_train_scale = scaler.fit_transform(feature_train)
fea_test_scale = scaler.fit_transform(feature_test)

def from_(y_train, y_test):
    
    C = np.logspace(-3,6,num=10)
    G = 0.1*np.arange(1,21,1)
    G_range = [round(i,1) for i in G]
    tuned_parameters = [{'kernel': ['rbf'], 'gamma': G_range, 'C': C}]
    
    grid_search = GridSearchCV(SVC(), tuned_parameters, cv=10,
                                                           scoring='accuracy', refit='accuracy')
    grid_search.fit(feature_train, y_train)
    best_parameters = grid_search.best_estimator_.get_params() 
    
    y_train_pred = grid_search.predict(feature_train)
    y_test_pred = grid_search.predict(feature_test)
   
    tmp_c = 0
    tmp_g = 0
    for para, val in list(best_parameters.items()):    
        print(para, val)
        if para == 'C':
            tmp_c = val
        if para == 'gamma':
            tmp_g = val
        else:
            continue
    return( tmp_c, tmp_g)


c_ = []
g_ = []
for category in categories:
    print(category)
    y_trian = label_train[category]
    y_test = label_test[category]
    
    tmp_c, tmp_g = from_(y_trian,y_test)
    c_.append(tmp_c)
    g_.append(tmp_g)
    
    

Family
C 10.0
cache_size 200
class_weight None
coef0 0.0
decision_function_shape ovr
degree 3
gamma 1.9
kernel rbf
max_iter -1
probability False
random_state None
shrinking True
tol 0.001
verbose False
Genus
C 100.0
cache_size 200
class_weight None
coef0 0.0
decision_function_shape ovr
degree 3
gamma 0.8
kernel rbf
max_iter -1
probability False
random_state None
shrinking True
tol 0.001
verbose False
Species
C 10.0
cache_size 200
class_weight None
coef0 0.0
decision_function_shape ovr
degree 3
gamma 1.7
kernel rbf
max_iter -1
probability False
random_state None
shrinking True
tol 0.001
verbose False


In [55]:
gau_ = pd.DataFrame({"class":['family', 'genus','species'],"C": c_, "Gamma": g_})
gau_

class      C  Gamma
0   family   10.0    1.9
1    genus  100.0    0.8
2  species   10.0    1.7

#### find exact match & hamming loss

In [56]:
#0
c_0 = SVC(C=c_[0] , gamma=g_[0], decision_function_shape='ovr',
                            kernel='rbf').fit(feature_train, label_train['Family'])

#1
c_1 = SVC(C=c_[1] , gamma=g_[1], decision_function_shape='ovr',
                            kernel='rbf').fit(feature_train, label_train['Genus'])

#2
c_2 = SVC(C=c_[2] , gamma=g_[2], decision_function_shape='ovr',
                            kernel='rbf').fit(feature_train, label_train['Species'])

y_pred = np.concatenate((np.expand_dims(c_0.predict(feature_test), axis=1), 
                                                 np.expand_dims(c_1.predict(feature_test), axis=1),
                                                 np.expand_dims(c_2.predict(feature_test), axis=1)),
                                                 axis=1)
y_pred=pd.DataFrame(y_pred)

y_true = np.concatenate((np.expand_dims(np.asarray(label_test['Family']), axis=1), 
                                       np.expand_dims(np.asarray(label_test['Genus']), axis=1),
                                       np.expand_dims(np.asarray(label_test['Species']), axis=1)), axis=1)
y_true=pd.DataFrame(y_true)

#==============================================
EM_score = 0
count_Ham = 0

for i in range(0,3): #family, genus, species
    count_EM = 0
    for j in range(len(y_pred[i])):
        if y_pred[i][j] == y_true[i][j]:
            count_EM +=1
        else:
            count_Ham +=1

EM_score = count_EM / y_true.shape[0]
Hamming_loss = count_Ham / (y_true.shape[0] * y_true.shape[1])

print("SVM with Gaussian Kernel with no Standard :")
print("Exact match score is:", EM_score)
print("Hamming Loss is:", Hamming_loss)

SVM with Gaussian Kernel with no Standard :
Exact match score is: 0.9911996294580825
Hamming Loss is: 0.008337193144974525


### SVM using Gaussian Kernel /with Standard

#### find c 

In [38]:
def from_svm(y_train, y_test):
    
    C = np.logspace(-3,6,num=10)
    G = 0.1*np.arange(1,21,1)
    tuned_parameters = [{'kernel': ['rbf'], 'gamma': G, 'C': C}]
    
    grid_search = GridSearchCV(SVC(), tuned_parameters, cv=10,
                                                           scoring='accuracy', refit='accuracy')
    grid_search.fit(fea_train_scale, y_train)
    best_parameters = grid_search.best_estimator_.get_params() 
    
    y_train_pred = grid_search.predict(fea_train_scale)
    y_test_pred = grid_search.predict(fea_test_scale)
   
    tmp_c = 0
    tmp_g = 0
    for para, val in list(best_parameters.items()):    
        print(para, val)
        if para == 'C':
            tmp_c = val
        if para == 'gamma':
            tmp_g = val
        else:
            continue
    return( tmp_c, tmp_g)


c_total = []
g_total = []
for category in categories:
    print(category)
    y_trian = label_train[category]
    y_test = label_test[category]
    
    tmp_c, tmp_g = from_svm(y_trian,y_test)
    c_total.append(tmp_c)
    g_total.append(tmp_g)
    

Family
C 10.0
cache_size 200
class_weight None
coef0 0.0
decision_function_shape ovr
degree 3
gamma 0.1
kernel rbf
max_iter -1
probability False
random_state None
shrinking True
tol 0.001
verbose False
Genus
C 10.0
cache_size 200
class_weight None
coef0 0.0
decision_function_shape ovr
degree 3
gamma 0.1
kernel rbf
max_iter -1
probability False
random_state None
shrinking True
tol 0.001
verbose False
Species
C 10.0
cache_size 200
class_weight None
coef0 0.0
decision_function_shape ovr
degree 3
gamma 0.1
kernel rbf
max_iter -1
probability False
random_state None
shrinking True
tol 0.001
verbose False


In [41]:
gau_val = pd.DataFrame({"class":['family', 'genus','species'],"C": c_total, "Gamma": g_total})
gau_val

class     C  Gamma
0   family  10.0    0.1
1    genus  10.0    0.1
2  species  10.0    0.1

#### find exact match & hamming loss 

In [25]:
#0
classif0 = SVC(C=c_total[0] , gamma=g_total[0], decision_function_shape='ovr',
                            kernel='rbf').fit(feature_train, label_train['Family'])

#1
classif1 = SVC(C=c_total[1] , gamma=g_total[1], decision_function_shape='ovr',
                            kernel='rbf').fit(feature_train, label_train['Genus'])

#2
classif2 = SVC(C=c_total[2] , gamma=g_total[2], decision_function_shape='ovr',
                            kernel='rbf').fit(feature_train, label_train['Species'])

y_pred = np.concatenate((np.expand_dims(classif0.predict(feature_test), axis=1), 
                                                 np.expand_dims(classif1.predict(feature_test), axis=1),
                                                 np.expand_dims(classif2.predict(feature_test), axis=1)),
                                                 axis=1)
y_pred=pd.DataFrame(y_pred)

y_true = np.concatenate((np.expand_dims(np.asarray(label_test['Family']), axis=1), 
                                       np.expand_dims(np.asarray(label_test['Genus']), axis=1),
                                       np.expand_dims(np.asarray(label_test['Species']), axis=1)), axis=1)
y_true=pd.DataFrame(y_true)

#==============================================
EM_score = 0
count_Ham = 0

for i in range(0,3): #family, genus, species
    count_EM = 0
    for j in range(len(y_pred[i])):
        if y_pred[i][j] == y_true[i][j]:
            count_EM +=1
        else:
            count_Ham +=1

EM_score = count_EM / y_true.shape[0]
Hamming_loss = count_Ham / (y_true.shape[0] * y_true.shape[1])

print("SVM with Gaussian Kernel:")
print("Exact match score is:", EM_score)
print("Hamming Loss is:", Hamming_loss)

SVM with Gaussian Kernel:
Exact match score is: 0.9759147753589624
Hamming Loss is: 0.02871699861046781


###  iii. L1 penal with SVM with standardized

#### find c:

In [6]:
from sklearn.svm import LinearSVC
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import StratifiedShuffleSplit
#train
feature_train = train_df.iloc[:,:22]
label_train = train_df.iloc[:,-4:-1]

#test
feature_test = test_df.iloc[:,:22]
label_test = test_df.iloc[:, -4:-1]
#---------------------------------------------------------------------
scaler=StandardScaler()
fea_train_scale = scaler.fit_transform(feature_train)
fea_test_scale = scaler.fit_transform(feature_test)
#-----------------------------------------------------------------------------

def from_L1(y_train, y_test):

    c_range = np.logspace(-3, 6, num=10)
    param_grid = dict(C=c_range)
    cross_val = StratifiedShuffleSplit(n_splits=10, random_state=0)
    grid_search_LF = GridSearchCV(LinearSVC(penalty='l1', dual=False), 
                                                                             param_grid=param_grid, 
                                                                             cv=cross_val, 
                                                                             scoring='accuracy')
    grid_search_LF.fit(fea_train_scale, y_train)
    best_parameters = grid_search_LF.best_estimator_.get_params() 
   
    tmp_c = 0
    for para, val in list(best_parameters.items()):    
        print(para, val)
        if para == 'C':
            tmp_c = val
        else:
            continue
    
    return( tmp_c)


c_total_l1 = []
for category in categories:
    print(category)
    y_trian = label_train[category]
    y_test = label_test[category]
    
    tmp_c = from_L1(y_trian,y_test)
    c_total_l1.append(tmp_c)


Family
C 1.0
class_weight None
dual False
fit_intercept True
intercept_scaling 1
loss squared_hinge
max_iter 1000
multi_class ovr
penalty l1
random_state None
tol 0.0001
verbose 0
Genus
C 100.0
class_weight None
dual False
fit_intercept True
intercept_scaling 1
loss squared_hinge
max_iter 1000
multi_class ovr
penalty l1
random_state None
tol 0.0001
verbose 0
Species
C 1.0
class_weight None
dual False
fit_intercept True
intercept_scaling 1
loss squared_hinge
max_iter 1000
multi_class ovr
penalty l1
random_state None
tol 0.0001
verbose 0


In [7]:
gau_val_l = pd.DataFrame({"class":['family', 'genus','species'],"C": c_total_l1})
gau_val_l

class      C
0   family    1.0
1    genus  100.0
2  species    1.0

#### find exact match & hamming loss

In [34]:
from sklearn.svm import LinearSVC
fea_train_scale = scaler.fit_transform(feature_train)
fea_test_scale = scaler.fit_transform(feature_test)

class0_l1 = LinearSVC(penalty='l1', loss='squared_hinge', dual=False, 
                                          C=c_total_l1[0], multi_class='ovr').fit(fea_train_scale, label_train['Family'])

class1_l1 = LinearSVC(penalty='l1',loss='squared_hinge', dual=False, 
                                          C=c_total_l1[1], multi_class='ovr').fit(fea_train_scale, label_train['Genus'])

class2_l1 = LinearSVC(penalty='l1',loss='squared_hinge', dual=False, 
                                          C=c_total_l1[2], multi_class='ovr').fit(fea_train_scale, label_train['Species'])


y_pred2 = np.concatenate((np.expand_dims(class0_l1.predict(fea_test_scale), axis=1), 
                         np.expand_dims(class1_l1.predict(fea_test_scale), axis=1),
                         np.expand_dims(class2_l1.predict(fea_test_scale), axis=1)), axis=1)

y_true = np.concatenate((np.expand_dims(np.asarray(label_test['Family']), axis=1), 
                         np.expand_dims(np.asarray(label_test['Genus']), axis=1),
                         np.expand_dims(np.asarray(label_test['Species']), axis=1)), axis=1)

EMscore = 0
count_Ham = 0
for i in range(0,y_true.shape[0]):
    count_EM = 0
    for j in range(0,y_true.shape[1]):
        if y_true[i][j]==y_pred2[i][j]:
            count_EM += 1
        else:
            count_Ham += 1
    if count_EM==3:
        EMscore += 1
        
EMscore = EMscore / y_true.shape[0]
Hamming_Loss = count_Ham / (y_true.shape[0] * y_true.shape[1])
            
print("L1-penalized SVM:")
print("The Exact Match score is:", EMscore)
print("The Hamming Loss is:", Hamming_Loss)

L1-penalized SVM:
The Exact Match score is: 0.9115331171838814
The Hamming Loss is: 0.05419175544233441


### iv SMOTE with L1 penal SVM with standardized

#### find c 

In [53]:
from sklearn.svm import LinearSVC
from imblearn.over_sampling import SMOTE 
from sklearn.model_selection import StratifiedShuffleSplit
feature_train = train_df.iloc[:,:22]
label_train = train_df.iloc[:,-4:-1]
scaler=StandardScaler()
fea_train_scale = scaler.fit_transform(feature_train)
fea_test_scale = scaler.fit_transform(feature_test)

def from_smote(y_train, y_test):
    
    smt = SMOTE(kind='regular')
    X_train_res, y_train_res = smt.fit_sample(fea_train_scale, y_train)
    
    c_range = np.logspace(-3, 6, num=10)
    param_grid = dict(C=c_range)
    cross_val = StratifiedShuffleSplit(n_splits=10, random_state=0)
    grid_search_smote_LF = GridSearchCV(LinearSVC(penalty='l1', dual=False), 
                                                                             param_grid=param_grid, 
                                                                             cv=cross_val, 
                                                                             scoring='accuracy')
    
    grid_search_smote_LF.fit(X_train_res, y_train_res)
    
    best_parameters = grid_search_smote_LF.best_estimator_.get_params() 
   
    tmp_c = 0
    for para, val in list(best_parameters.items()):    
        print(para, val)
        if para == 'C':
            tmp_c = val
        else:
            continue
    
    return( tmp_c)

#====================================
c_total_s = []
for category in categories:
    print(category)
    y_trian = label_train[category]
    y_test = label_test[category]
    
    tmp_c = from_smote(y_trian,y_test)
    c_total_s.append(tmp_c)
    

Family
C 100.0
class_weight None
dual False
fit_intercept True
intercept_scaling 1
loss squared_hinge
max_iter 1000
multi_class ovr
penalty l1
random_state None
tol 0.0001
verbose 0
Genus
C 100.0
class_weight None
dual False
fit_intercept True
intercept_scaling 1
loss squared_hinge
max_iter 1000
multi_class ovr
penalty l1
random_state None
tol 0.0001
verbose 0
Species
C 100.0
class_weight None
dual False
fit_intercept True
intercept_scaling 1
loss squared_hinge
max_iter 1000
multi_class ovr
penalty l1
random_state None
tol 0.0001
verbose 0


In [54]:
gau_val_s = pd.DataFrame({"class":['family', 'genus','species'],"C": c_total_s})
gau_val_s

class      C
0   family  100.0
1    genus  100.0
2  species  100.0

#### find exact match & hamming loss

In [55]:
from sklearn.svm import LinearSVC
from imblearn.over_sampling import SMOTE 

sm = SMOTE(kind='regular')

# Classifier for label 'Family'
#c=1000
train_sm_0,train_l_sm_0 = sm.fit_sample(feature_train, label_train['Family'])
class0_smote = LinearSVC(penalty='l1', loss='squared_hinge', dual=False, 
                                                 C=c_total_s[0], multi_class='ovr')
class0_smote.fit(train_sm_0, train_l_sm_0)

#c=100
# Classifier for label 'Genus'
train_sm_1,train_l_sm_1 = sm.fit_sample(feature_train, label_train['Genus'])
class1_smote = LinearSVC(penalty='l1',loss='squared_hinge', dual=False, 
                                                 C=c_total_s[1], multi_class='ovr')
class1_smote.fit(train_sm_1,train_l_sm_1)

#c=1000000
# Classifier for label 'Species'
train_sm_2,train_l_sm_2 = sm.fit_sample(feature_train, label_train['Species'])
class2_smote = LinearSVC(penalty='l1',loss='squared_hinge', dual=False, 
                                                 C=c_total_s[2], multi_class='ovr')
class2_smote.fit(train_sm_2,train_l_sm_2)


y_pred3 = np.concatenate((np.expand_dims(class0_smote.predict(feature_test), axis=1), 
                                                    np.expand_dims(class1_smote.predict(feature_test), axis=1),
                                                    np.expand_dims(class2_smote.predict(feature_test), axis=1)), axis=1)

y_true = np.concatenate((np.expand_dims(np.asarray(label_test['Family']), axis=1), 
                                                np.expand_dims(np.asarray(label_test['Genus']), axis=1),
                                                np.expand_dims(np.asarray(label_test['Species']), axis=1)), axis=1)

EMscore = 0
count_Ham = 0
for i in range(0,y_true.shape[0]):
    count_EM = 0
    for j in range(0,y_true.shape[1]):
        if y_true[i][j]==y_pred3[i][j]:
            count_EM += 1
        else:
            count_Ham += 1
    if count_EM==3:
        EMscore += 1
        
EMscore = EMscore / y_true.shape[0]
Hamming_Loss = count_Ham / (y_true.shape[0] * y_true.shape[1])
            
print("SVM with SMOTE:")
print("The Exact Match score is:", EMscore)
print("The Hamming Loss is:", Hamming_Loss)

SVM with SMOTE:
The Exact Match score is: 0.8554886521537749
The Hamming Loss is: 0.07688744789254284


|type| The Exact Match| The Hamming Loss|
|:---|---|---:|
|SVM with Gaussian|0.9911996294580825|0.008337193144974525
|SVM with Gaussian with standardized|  0.9759147753589624|0.02871699861046781|
|SVM with L1 with standardized |0.9119962945808244| 0.05419175544233441|
|SVM with SMOTE with standardized|0.8554886521537749|0.07688744789254284|


#### Ans:
As the above table, we can see the results that:

To sum up, **the SVM with Gaussian** has the best Exact Match and the lowest Hamming Loss while **the SVM with Gaussian with standardized** is a little worest, which maybe contrary to the general theory. In this case, it may be possible that the data selected at random is not suitable for standardization

Because **Gaussian kernel** is able to fit for a non-linear relationship among classes, it has the best results. 

Also, **the SVM with L1-penal with standardized** and **the SVM with SMOTE with L1-penal with standardized** have not bad scores. 



# 2. K-Means Clustering on a Multi-Class and Multi-Label Data Set

### a. Use k-means clustering
#### Research:

**Hamming loss**
The fraction of the wrong labels to the total number of labels
<br>
<br>
**Hamming score**
The set of labels predicted for a sample must exactly match the corresponding set of labels in y_true.
<br>
<br>
**Hamming distance:**
Hamming distance is a metric for comparing two binary data strings. While comparing two binary strings of equal length, Hamming distance is the number of bit positions in which the two bits are different. (https://www.tutorialspoint.com/what-is-hamming-distance)
<br>
<br>
So, to calculate the distance, I compared two **2D-array (y_pred, y_true)**.  I seem 3-string-list [family, genus, species] as a unit.  Then, calculate the total number of incorrect in whole y_pred data set.  So, in some cases, the hamming distance may be similar to the hamming loss. 


## MonteCarlo Simulation 50 times

In [47]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn import metrics
from sklearn.cluster import KMeans
from sklearn.metrics import hamming_loss
import random
import statistics
from scipy.spatial import distance

# Loading the DATA
all_data = pd.read_csv("D:/552 Machine Learning for Data Science/homework/Homework5/Anuran Calls (MFCCs)/Frogs_MFCCs.csv")
all_data=all_data.drop('RecordID',axis=1)

cols=[e for e in all_data if e not in ( 'Family','Genus','Species')]
X=all_data.loc[:,cols]
Y=all_data.loc[:, ['Family','Genus','Species']]

total_k=[]
final_ham_dist=[]
final_ham_loss=[]
final_ham_score=[]

fam_maj_trip={p:[] for p in range(1,51)}
genus_maj_trip={p:[] for p in range(1,51)}
species_maj_trip={p:[] for p in range(1,51)}

for time in range(1,51):
    print("********         ", time,"        ********")
   
    # --------------------a. Finding the Optimal K value ------------------------
    '''
    https://stackoverflow.com/questions/51138686/how-to-use-silhouette-score-in-k-means-clustering-from-sklearn-library
    '''
    silh_avg = dict()
    for k in range(2,51):
        k_means = KMeans(n_clusters=k,init='k-means++',random_state=time).fit(X)
        labels = k_means.labels_
        silh_avg.update({k:(metrics.silhouette_score(X, labels))})
   #print("Average Silhoutte score values : ",silh_avg)
    
    optimal_k = max(silh_avg,key=silh_avg.get)
    total_k.append(optimal_k)
    K = optimal_k
    
    #-----------------------------------------------------------------------------
    tmp_X=X
    k_means_ = KMeans(n_clusters=K, random_state=time).fit(tmp_X)
    cluster_labels = k_means_.labels_
    
    clusters = pd.concat([tmp_X,Y,pd.DataFrame({'labels':cluster_labels.tolist()})],axis = 1)
    clusters['labels'].value_counts()

    
    #-----------------------  b.  determine majority labels ---------------------------
    for k in range(0,K):
        find= clusters[clusters['labels']==k]
        print('Cluster',k+1)
        print('Majority class in family - ',find['Family'].value_counts().index[0])
        print('Majority class in genus - ',find['Genus'].value_counts().index[0])
        print('Majority class in species - ',find['Species'].value_counts().index[0])

   
    #----------------------  Determine the Majority Triplet f -----------------
    '''
    http://rasbt.github.io/mlxtend/user_guide/classifier/EnsembleVoteClassifier/
    '''
    maj_trip = {k:[] for k in range(0,K)}
    for k in range(0,K):
        c_value = clusters[clusters['labels']==k]
        maj_trip[k].append(c_value['Family'].value_counts().index[0])
        maj_trip[k].append(c_value['Genus'].value_counts().index[0])
        maj_trip[k].append(c_value['Species'].value_counts().index[0])
        
        fam_maj_trip[time].append(c_value['Family'].value_counts().index[0])
        genus_maj_trip[time].append(c_value['Genus'].value_counts().index[0])
        species_maj_trip[time].append(c_value['Species'].value_counts().index[0])

    clusters['family_pred'] = 'none'
    clusters['genus_pred'] = 'none'
    clusters['species_pred'] = 'none'

    for k in range(0,K): 
        clusters['family_pred'] = np.where(clusters['labels']==k,maj_trip[k][0],clusters['family_pred'])
        clusters['genus_pred'] = np.where(clusters['labels']==k,maj_trip[k][1],clusters['genus_pred'])
        clusters['species_pred'] = np.where(clusters['labels']==k,maj_trip[k][2],clusters['species_pred'])
        

    #---------------------------   c  -----------------------------------------
    #---------------  Average hamming distance  -------------------------
    y_true = Y
    y_pred = clusters[['family_pred','genus_pred','species_pred']]
    
    tmp_mean = []
    for i in range(y_true.shape[0]):
        tmp = distance.hamming(y_true.loc[i],y_pred.loc[i])
        tmp_m= np.mean(tmp)
        tmp_mean.append(tmp_m)

    ham_dist = np.mean(tmp_mean)
    print("Hamming Distance: ", ham_dist)
    
    # ----------------   Average Hamming Score  ------------------------
    
    fam_s=hamming_loss(clusters['Family'],clusters['family_pred'])
    gen_s=hamming_loss(clusters['Genus'],clusters['genus_pred'])
    spec_s=hamming_loss(clusters['Species'],clusters['species_pred'])

    ham_loss_s=(fam_s+gen_s+spec_s)/3
    
    print("Hamming Loss : ",ham_loss_s)
    print("Hamming Score : ",1-ham_loss_s)
    print("\n")
    
    # ----------------  average & standard deviation -----------------------
    final_ham_dist.append(ham_dist)
    final_ham_loss.append(ham_loss_s)
    final_ham_score.append((1-ham_loss_s))
    
    
tmp_major=list(zip(list(fam_maj_trip.values()),
                                     list(genus_maj_trip.values()),
                                     list(species_maj_trip.values())))

major_trip={}
for i in range(0,50):
    major_trip.update({i:list(zip(tmp_major[i][0],tmp_major[i][1],tmp_major[i][2]))})

time_=range(1,51)
result=pd.DataFrame({"Iteration" : time_,
                                          "Optimal K":total_k,
                                          "Hamming Distance":final_ham_dist, 
                                          "Hamming Loss":final_ham_loss, 
                                          "Hamming Score":final_ham_score,
                                          "Majority labels (Family,Genus,Species)":list(major_trip.values())})

********          1         ********
Cluster 1
Majority class in family -  Hylidae
Majority class in genus -  Hypsiboas
Majority class in species -  HypsiboasCinerascens
Cluster 2
Majority class in family -  Dendrobatidae
Majority class in genus -  Ameerega
Majority class in species -  Ameeregatrivittata
Cluster 3
Majority class in family -  Leptodactylidae
Majority class in genus -  Adenomera
Majority class in species -  AdenomeraHylaedactylus
Cluster 4
Majority class in family -  Hylidae
Majority class in genus -  Hypsiboas
Majority class in species -  HypsiboasCordobae
Hamming Distance:  0.22242297892054666
Hamming Loss :  0.22242297892054666
Hamming Score :  0.7775770210794534


********          2         ********
Cluster 1
Majority class in family -  Hylidae
Majority class in genus -  Hypsiboas
Majority class in species -  HypsiboasCordobae
Cluster 2
Majority class in family -  Leptodactylidae
Majority class in genus -  Adenomera
Majority class in species -  AdenomeraHylaedactylu

In [48]:
print("|Average Hamming Distance|", np.mean(final_ham_dist))
print("|Standard Deviation of Hamming Distance|", np.std(final_ham_dist))

print("|Average Hamming Loss|", np.mean(final_ham_loss))
print("|Standard Deviation of Hamming Loss|", np.std(final_ham_loss))

print("|Average Hamming Score|", np.mean(final_ham_score))
print("|Standard Deviation of Hamming Score|", np.std(final_ham_score))


|Average Hamming Distance| 0.2237535325457493
|Standard Deviation of Hamming Distance| 0.010335099703838932
|Average Hamming Loss| 0.2237535325457493
|Standard Deviation of Hamming Loss| 0.010335099703838927
|Average Hamming Score| 0.7762464674542505
|Standard Deviation of Hamming Score| 0.010335099703838922


## b. Majority labels of each cluster

In [49]:
fam_result=pd.DataFrame(fam_maj_trip)
fam_result

1                2                3                4   \
0          Hylidae          Hylidae          Hylidae  Leptodactylidae   
1    Dendrobatidae  Leptodactylidae          Hylidae          Hylidae   
2  Leptodactylidae          Hylidae  Leptodactylidae          Hylidae   
3          Hylidae  Leptodactylidae    Dendrobatidae    Dendrobatidae   

                5                6                7                8   \
0    Dendrobatidae  Leptodactylidae  Leptodactylidae  Leptodactylidae   
1          Hylidae          Hylidae          Hylidae    Dendrobatidae   
2          Hylidae          Hylidae    Dendrobatidae          Hylidae   
3  Leptodactylidae    Dendrobatidae          Hylidae          Hylidae   

                9                10  ...               41               42  \
0          Hylidae  Leptodactylidae  ...  Leptodactylidae  Leptodactylidae   
1  Leptodactylidae          Hylidae  ...          Hylidae          Hylidae   
2    Dendrobatidae    Dendrobatidae  ...          Hylidae          Hylidae   
3          Hylidae          Hylidae  ...    Dendrobatidae    Dendrobatidae   

                43               44               45               46  \
0  Leptodactylidae  Leptodactylidae          Hylidae          Hylidae   
1          Hylidae          Hylidae  Leptodactylidae  Leptodactylidae   
2          Hylidae    Dendrobatidae    Dendrobatidae    Dendrobatidae   
3    Dendrobatidae          Hylidae          Hylidae          Hylidae   

                47               48               49               50  
0          Hylidae  Leptodactylidae  Leptodactylidae  Leptodactylidae  
1  Leptodactylidae          Hylidae          Hylidae    Dendrobatidae  
2          Hylidae          Hylidae    Dendrobatidae          Hylidae  
3    Dendrobatidae          Hylidae          Hylidae          Hylidae  

[4 rows x 50 columns]

In [50]:
genus_result = pd.DataFrame(genus_maj_trip)
genus_result

1          2          3          4          5          6   \
0  Hypsiboas  Hypsiboas  Hypsiboas  Adenomera   Ameerega  Adenomera   
1   Ameerega  Adenomera  Hypsiboas  Hypsiboas  Hypsiboas  Hypsiboas   
2  Adenomera  Hypsiboas  Adenomera  Hypsiboas  Hypsiboas  Hypsiboas   
3  Hypsiboas  Adenomera   Ameerega   Ameerega  Adenomera   Ameerega   

          7          8          9          10  ...         41         42  \
0  Adenomera  Adenomera  Hypsiboas  Adenomera  ...  Adenomera  Adenomera   
1  Hypsiboas   Ameerega  Adenomera  Hypsiboas  ...  Hypsiboas  Hypsiboas   
2   Ameerega  Hypsiboas   Ameerega   Ameerega  ...  Hypsiboas  Hypsiboas   
3  Hypsiboas  Hypsiboas  Hypsiboas  Hypsiboas  ...   Ameerega   Ameerega   

          43         44         45         46         47         48  \
0  Adenomera  Adenomera  Hypsiboas  Hypsiboas  Hypsiboas  Adenomera   
1  Hypsiboas  Hypsiboas  Adenomera  Adenomera  Adenomera  Hypsiboas   
2  Hypsiboas   Ameerega   Ameerega   Ameerega  Hypsiboas  Hypsiboas   
3   Ameerega  Hypsiboas  Hypsiboas  Hypsiboas   Ameerega  Hypsiboas   

          49         50  
0  Adenomera  Adenomera  
1  Hypsiboas   Ameerega  
2   Ameerega  Hypsiboas  
3  Hypsiboas  Hypsiboas  

[4 rows x 50 columns]

In [51]:
species_result = pd.DataFrame(species_maj_trip)
species_result

1                       2                       3   \
0    HypsiboasCinerascens       HypsiboasCordobae       HypsiboasCordobae   
1      Ameeregatrivittata  AdenomeraHylaedactylus    HypsiboasCinerascens   
2  AdenomeraHylaedactylus       HypsiboasCordobae  AdenomeraHylaedactylus   
3       HypsiboasCordobae          AdenomeraAndre      Ameeregatrivittata   

                       4                       5                       6   \
0  AdenomeraHylaedactylus      Ameeregatrivittata  AdenomeraHylaedactylus   
1       HypsiboasCordobae       HypsiboasCordobae       HypsiboasCordobae   
2    HypsiboasCinerascens    HypsiboasCinerascens    HypsiboasCinerascens   
3      Ameeregatrivittata  AdenomeraHylaedactylus      Ameeregatrivittata   

                       7                       8                       9   \
0  AdenomeraHylaedactylus  AdenomeraHylaedactylus       HypsiboasCordobae   
1       HypsiboasCordobae      Ameeregatrivittata  AdenomeraHylaedactylus   
2      Ameeregatrivittata       HypsiboasCordobae      Ameeregatrivittata   
3    HypsiboasCinerascens    HypsiboasCinerascens    HypsiboasCinerascens   

                       10  ...                      41  \
0  AdenomeraHylaedactylus  ...  AdenomeraHylaedactylus   
1       HypsiboasCordobae  ...       HypsiboasCordobae   
2      Ameeregatrivittata  ...    HypsiboasCinerascens   
3    HypsiboasCinerascens  ...      Ameeregatrivittata   

                       42                      43                      44  \
0  AdenomeraHylaedactylus  AdenomeraHylaedactylus  AdenomeraHylaedactylus   
1       HypsiboasCordobae    HypsiboasCinerascens       HypsiboasCordobae   
2    HypsiboasCinerascens       HypsiboasCordobae      Ameeregatrivittata   
3      Ameeregatrivittata      Ameeregatrivittata    HypsiboasCinerascens   

                       45                      46                      47  \
0       HypsiboasCordobae       HypsiboasCordobae       HypsiboasCordobae   
1  AdenomeraHylaedactylus  AdenomeraHylaedactylus  AdenomeraHylaedactylus   
2      Ameeregatrivittata      Ameeregatrivittata    HypsiboasCinerascens   
3    HypsiboasCinerascens    HypsiboasCinerascens      Ameeregatrivittata   

                       48                      49                      50  
0  AdenomeraHylaedactylus  AdenomeraHylaedactylus  AdenomeraHylaedactylus  
1       HypsiboasCordobae    HypsiboasCinerascens      Ameeregatrivittata  
2    HypsiboasCinerascens      Ameeregatrivittata    HypsiboasCinerascens  
3       HypsiboasCordobae       HypsiboasCordobae       HypsiboasCordobae  

[4 rows x 50 columns]

In [52]:
result

Iteration  Optimal K  Hamming Distance  Hamming Loss  Hamming Score  \
0           1          4          0.222423      0.222423       0.777577   
1           2          4          0.245263      0.245263       0.754737   
2           3          4          0.222423      0.222423       0.777577   
3           4          4          0.222423      0.222423       0.777577   
4           5          4          0.222423      0.222423       0.777577   
5           6          4          0.222423      0.222423       0.777577   
6           7          4          0.222423      0.222423       0.777577   
7           8          4          0.222423      0.222423       0.777577   
8           9          4          0.222423      0.222423       0.777577   
9          10          4          0.222423      0.222423       0.777577   
10         11          4          0.222423      0.222423       0.777577   
11         12          4          0.222423      0.222423       0.777577   
12         13          4          0.222423      0.222423       0.777577   
13         14          4          0.222423      0.222423       0.777577   
14         15          4          0.222469      0.222469       0.777531   
15         16          4          0.222423      0.222423       0.777577   
16         17          4          0.233727      0.233727       0.766273   
17         18          4          0.222423      0.222423       0.777577   
18         19          4          0.221774      0.221774       0.778226   
19         20          4          0.234051      0.234051       0.765949   
20         21          4          0.222423      0.222423       0.777577   
21         22          4          0.222423      0.222423       0.777577   
22         23          4          0.222469      0.222469       0.777531   
23         24          4          0.222423      0.222423       0.777577   
24         25          4          0.222423      0.222423       0.777577   
25         26          4          0.222423      0.222423       0.777577   
26         27          4          0.222469      0.222469       0.777531   
27         28          4          0.222423      0.222423       0.777577   
28         29          4          0.222423      0.222423       0.777577   
29         30          4          0.222423      0.222423       0.777577   
30         31          4          0.222423      0.222423       0.777577   
31         32          4          0.222423      0.222423       0.777577   
32         33          4          0.222423      0.222423       0.777577   
33         34          4          0.186055      0.186055       0.813945   
34         35          4          0.222423      0.222423       0.777577   
35         36          4          0.222423      0.222423       0.777577   
36         37          4          0.222423      0.222423       0.777577   
37         38          4          0.222423      0.222423       0.777577   
38         39          4          0.222423      0.222423       0.777577   
39         40          4          0.222423      0.222423       0.777577   
40         41          4          0.222423      0.222423       0.777577   
41         42          4          0.222423      0.222423       0.777577   
42         43          4          0.222423      0.222423       0.777577   
43         44          4          0.222423      0.222423       0.777577   
44         45          4          0.222423      0.222423       0.777577   
45         46          4          0.222423      0.222423       0.777577   
46         47          4          0.222423      0.222423       0.777577   
47         48          4          0.280056      0.280056       0.719944   
48         49          4          0.222423      0.222423       0.777577   
49         50          4          0.222423      0.222423       0.777577   

               Majority labels (Family,Genus,Species)  
0   [(Hylidae, Hypsiboas, HypsiboasCinerascens), (...  
1   [(Hylidae, Hypsiboas, HypsiboasCordobae), (Lep...  
2   [(Hyli

## c. the average and the standard deviation of Hamming distance, Hamming score, and Hamming loss

|type|result|
|:---|---:|
|Average Hamming Distance| 0.2237535325457493
|Standard Deviation of Hamming Distance| 0.010335099703838932
|Average Hamming Loss| 0.2237535325457493
|Standard Deviation of Hamming Loss| 0.010335099703838927
|Average Hamming Score| 0.7762464674542505
|Standard Deviation of Hamming Score| 0.010335099703838922

# ISLR 10.7.2 

#### Ans: Peiying_Lyu_ISLR10.7.2.pdf